fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/search/LuceneOptimizedIndexSearcher.java

```diff
@@ -115,44 +139,12 @@ public <C extends Collector, T> T search(Query query, CollectorManager<C, T> col
             final Weight weight = createWeight(query, scoreMode, 1);
             final List<Future<C>> topDocsFutures = new ArrayList<>(leafSlices.length);
 
-            for (int i = 0; i < leafSlices.length - 1; ++i) {
+            for (int i = 0; i < leafSlices.length; ++i) {
                 final LeafReaderContext[] leaves = leafSlices[i].leaves;
                 final C collector = collectors.get(i);
-
-                // 1. spawn a list of parallel tasks that interact with the DB for better throughput.
-                final List<CompletableFuture<Pair<LeafReaderContext, Pair<LeafCollector, BulkScorer>>>> dependencies = Arrays.stream(leaves).map(ctx -> CompletableFuture.supplyAsync(() -> {
-                    try {
-                        final var result = Pair.of(collector.getLeafCollector(ctx), weight.bulkScorer(ctx));
-                        return Pair.of(ctx, result);
-                    } catch (IOException e) {
-                        return null;
-                    }
-                }, getExecutor())).collect(Collectors.toList());
-
-                // 2. once we finish processing all of them, we can proceed with the final scoring task.
-                final var future = CompletableFuture.allOf(dependencies.toArray(new CompletableFuture[dependencies.size()])).thenApplyAsync(ignored -> {
-                    dependencies.stream().map(CompletableFuture::join).collect(Collectors.toList())
-                            .forEach((Pair<LeafReaderContext, Pair<LeafCollector, BulkScorer>> result) -> {
-                                final Pair<LeafCollector, BulkScorer> scorer = result.getRight();
-                                final LeafReaderContext ctx = result.getLeft();
-                                if (scorer != null && scorer.getLeft() != null && scorer.getRight() != null) {
-                                    try {
-                                        scorer.getRight().score(scorer.getLeft(), ctx.reader().getLiveDocs());
-                                    } catch (IOException e) {
-                                        // no-op just ignore.
-                                    }
-                                }
-                            });
-                    return collector;
-                }, getExecutor());
+                final CompletableFuture<C> future = searchOptimized(executor, weight, Arrays.asList(leaves), collector);
                 topDocsFutures.add(future);
             }
-
-            final LeafReaderContext[] leaves = leafSlices[leafSlices.length - 1].leaves;
-            final C collector = collectors.get(leafSlices.length - 1);
-            // execute the last on the caller thread
-            search(Arrays.asList(leaves), weight, collector);
-            topDocsFutures.add(CompletableFuture.completedFuture(collector));
```

Comment:
This was inherited from Lucene's implementation, but I find it more consistent to delegate work to the desginated thread pool. What's your thoughts on this @MMcM?

Reply:
I thought the same thing. It seems silly to not just make a consistent set of futures. Who cares if the calling thread blocks on them all instead of (first) doing one of them? On the other hand, I do see the benefit of (mostly) copying the code for future maintainability.